pkg/kubelet/images/image_manager.go

```diff
@@ -63,30 +94,73 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 	} else {
 		puller = newParallelImagePuller(imageService, maxParallelImagePulls)
 	}
+
+	stateImpl, err := NewCheckpointState(stateFileDirectory, secretPulledImageFileName)
+	// TODO (pacoxu) make a Noop state implementation for test
+	if err != nil {
+		klog.ErrorS(err, "Could not initialize checkpoint manager.")
+	}
+
 	return &imageManager{
 		recorder:               recorder,
 		imageService:           imageService,
 		backOff:                imageBackOff,
 		puller:                 puller,
 		podPullingTimeRecorder: podPullingTimeRecorder,
+		keyring:                keyring,
+		state:                  stateImpl,
 	}
 }
 
 // shouldPullImage returns whether we should pull an image according to
 // the presence and pull policy of the image.
-func shouldPullImage(container *v1.Container, imagePresent bool) bool {
+func shouldPullImage(container *v1.Container, imagePresent, pulledBySecret, ensuredBySecret bool) bool {
 	if container.ImagePullPolicy == v1.PullNever {
 		return false
 	}
 
-	if container.ImagePullPolicy == v1.PullAlways ||
-		(container.ImagePullPolicy == v1.PullIfNotPresent && (!imagePresent)) {
-		return true
+	if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) { // ungated old behavior
+		if container.ImagePullPolicy == v1.PullAlways ||
+			(container.ImagePullPolicy == v1.PullIfNotPresent && (!imagePresent)) {
+			return true
+		}
+	} else { // new gated behavior
```

Comment:
nit: I'm personally more of a fan of more return statements rather than having a bunch of nested elses. up to you

Reply:
I'd prefer to use a switch on the gate.. was just trying to make it easier for reviewers to read the policy change

Reply:
refactor the func code in https://github.com/kubernetes/kubernetes/pull/114847/commits/28a55af28703b0d8c465c619673456b5bcba0a95 to make it more clear.

Use `switch` for `ImagePullPolicy` and return early in the below logic.